### ISSUE TITLE
Winforms like designer with optional hot reload xaml window

### ISSUE THREAD

**Issue by Shadowblitz16** [dotnet/maui#3439]
### Description

Can we have a winforms like designer for maui for vs and vs code?
Winforms has one of the best drag and drop designers that Microsoft has ever produced.
Only being compared to something like vb6.

Of course we would need a live xaml editor with hot reload but both the xaml editor and the forms designer should be closable for small single monitor work spaces. 

### Public API Changes

It doesn't really need any public api changes as far as I know.
Its more of a visual designer suggestion.

### Intended Use-Case

Rapid prototyping and development for both xaml and winform developers .

**Comment by XeonG** [comment]
"Winforms has one of the best drag and drop designers that Microsoft has ever produced."

That was clearly from a day and time when MS had employees who could actually make decent tools and put together consistent UI's.

Instead they buy crap like Xamarin. And there OS looks like it's been pulled apart and taken over by literal retards turning into absolute rubbish.. If it wasn't for the third party dev community putting up patches to fix it, like Explorer Patcher https://github.com/valinet/ExplorerPatcher I would have already left development on this  trainwreck that is Windows.

And I dunno who made there Dark Mode theme for Win11 but it was like they hired a bunch of interns who just thought rounding some corners off was enough.. Leaving just about every winform GUI control completely unthemed while breaking support for anyone to fix it. So most UI  you see in Windows is just white.. in a dark mode theme.. moron development.

Anyway I'll just stick to using JavaFX they have a Scenebuilder for quick prototyping and the build pipelines work for all platforms..ie Linux included..  Or use QT 

MS direction with just about everything now is hair brained and inconsistent, about the consistent thing is it being constantly broken, barely fit for purpose and really  unfinished rubbish.

**Comment by MagicAndre1981** [comment]
I still prefer a winforms-like drag-and-drop designer, I hate this XAML only "design" way. I spend most time trying to make the UI look somehow useable. And the hotreload thing is an ugly workaround which never really works for me.

**Comment by pieckenst** [comment]
Yeah a xaml drag and drop designer would be a thing good to have , and make designing application ui simpler and more intuitive

**Comment by davepruitt** [comment]
Similar issues continue to get raised, but Microsoft keeps closing them:

https://github.com/dotnet/maui/issues/6782
https://github.com/dotnet/maui/issues/8610

**Comment by GuildOfCalamity** [comment]
How the hell does WinForms, WPF & UWP have a real-time designer and MAUI somehow does not?????
Is there no QA process at MS anymore? How could you ship VS2022 and not make sure this works... the designer was kind of buggy in VS2019 but I thought it would be completely improved by VS2022.

**Comment by dazlab** [comment]
Removing the XAML Designer and telling people to use Hot Reload is a major fail. You can only make minor changes before having to rebuild and re-deploy in Hot Reload. It's an absolute nightmare. In the Out-of-the-Box default MAUI app I tried adding an extra button and it wanted a rebuild, which took over 1 minute. This is your solution? Having to do this every time you make a small tweak to the design in not productive. 

Tools like Basic4Android allow you to connect to a device and move elements ON THE DEVICE during the design phase, and has had this for years. The Xamarin XAML Designer was fairly buggy, true, but this is a huge step backwards. 

Why, oh why, do M$ keep making these schoolboy errors. One step forward and three back.

**Comment by ShiiikK** [comment]
Hope MAUI is going to have this again.
Otherwise i cant understand the big step back.

I think Flutterflow is going to show us how this can work.
Also WinForms or WPF show us how big this improvement is.

Please bring us a Designer, Hot Reload cant replace this.

BRING US THE GREAT SH*T !

**Comment by halal-beef** [comment]
yeah, just loaded up maui wanted to make a decent ui and im met with xaml text, i like to look at things visually... to you know actually design

**Comment by saber-wang** [comment]
Without a designer, it is difficult to write the desired interface.

**Comment by jaworekplay** [comment]
Personally I don't care for WinForms designer, since it is using Margins (yuck!) when placing elements on a control. Preview of Window/UserControl would be great. Some apps are huge and it takes a long time to load, HotReload also won't really work if some assemblies are being loaded manually. Really a step backwards. No need for drag and drop.
#KeepDesignerSimple

**Comment by davepruitt** [comment]
> Personally I don't care for WinForms designer, since it is using Margins (yuck!) when placing elements on a control. Preview of Window/UserControl would be great. Some apps are huge and it takes a long time to load, HotReload also won't really work if some assemblies are being loaded manually. Really a step backwards. No need for drag and drop. #KeepDesignerSimple

Yeah most people aren't wanting a drag-and-drop designer, but rather a XAML designer/previewer similar to what is already available in WPF, or what was previously available in Xamarin Forms.

**Comment by Siox0911** [comment]
Plz vote up this on the feedback page.
https://developercommunity.visualstudio.com/t/MAUI-XAML-Designer-and-editor-not-workin/1589956

**Comment by k3ldar** [comment]
Coming from a WinForm/Web background (no WPF etc) I really wanted to try MAUI with a new project I have, however, without some form of gui editor I have no starting reference, especially as intellisense does not show anything and trying to obtain help from within the .xaml loads a generic page (showing a gui editor).  

With no reference of what controls or properties are  available within MAUI some form of gui editor is sort of essential from my PoV.

Hot reload is great, but only if you know what controls to add to a xaml file and what properties to set, so a +1 from me on this feature request

**Comment by Khaldoun-Baz** [comment]
I don't know why Microsoft is always complicating things, it is that hard to make a visual designer or previewer for MAUI?
Really I do not understand it's like pushing developers away, I do not know if I will invest more time in MAUI.

**Comment by dragnilar** [comment]
I'm going to have to add to the demand for a designer, or at least some type of in vs preview. Hot reload is too unreliable and too slow to get a visual representation of xaml changes.

**Comment by vigouredelaruse** [comment]
if this is a voting situation i vote for maui support from the xaml designer. with one product announcement much discontent would be eliminated

also
* private enterprise is not illegal everywhere
* where private enterprises are legal nothing prevents private enterprises from publishing a xaml designer on github, or the app store

**Comment by da-vics** [comment]
> this is a voting situation i vote for m



> > Personally I don't care for WinForms designer, since it is using Margins (yuck!) when placing elements on a control. Preview of Window/UserControl would be great. Some apps are huge and it takes a long time to load, HotReload also won't really work if some assemblies are being loaded manually. Really a step backwards. No need for drag and drop. #KeepDesignerSimple
> 
> Yeah most people aren't wanting a drag-and-drop designer, but rather a XAML designer/previewer similar to what is already available in WPF, or what was previously available in Xamarin Forms.

100%

**Comment by zbyna** [comment]
From my point of view XALM is horrible even for using not mentioned large projects debugging. It is better to use https://docs.godotengine.org/en/stable/tutorials/ui/index.html for creating UI than recent ms products. There are good to use multiplatform  RAD tools like: https://github.com/wxFormBuilder/wxFormBuilder or  https://www.lazarus-ide.org/
but not for c#.

**Comment by Siox0911** [comment]
Fun fact. You can design a .Net maui App in Blender... without a XAML Designer. 😂

**Comment by MisinformedDNA** [comment]
> Fun fact. You can design a .Net maui App in Blender...

How?

**Comment by dragnilar** [comment]
> > Fun fact. You can design a .Net maui App in Blender...
> 
> How?

Same question. I tried looking into this and I could not find anything.

**Comment by GuildOfCalamity** [comment]
> > > Fun fact. You can design a .Net maui App in Blender...
> > 
> > How?
> 
> Same question. I tried looking into this and I could not find anything.

That's because it's a vapid comment, he's saying you can create a graphic visual. You can also draw a circle in **Paint**, but that's not going to help anyone with the lack of a real-time designer in VS.

**Comment by vigouredelaruse** [comment]
during the review of these githubs for the product roadmap one hastens to remind the reviewers that some of the users in the market have never taken a design course in our lives, but still have to mock user interfaces while we're building back ends, or trying out product ideas to bring to management

**Comment by davepruitt** [comment]
This thread currently ranks 4th in total "likes", 2nd in total "hearts", and 17th in total comments compared to all other open issues on this repository. A similar issue located in the WinUI repository ranks 2nd in total comments and 3rd for most likes for that repository.

Yet for some reason Microsoft still has not fully acknowledged their failure and hasn't given any solid evidence that a XAML designer is in the works. This needs to happen for MAUI, WinUI, and yes it even needs to happen retroactively for Xamarin Forms - since it already existed for Xamarin Forms and they took it away.

**Comment by dragnilar** [comment]
@davepruitt  I agree.

IF they are working on anything, I suspect they won't show a hint of it until it's ready. MS is tight lipped about what goes on with VS developments; for example - unless there was some talk of it that I missed, they kept us all pretty much in the dark about the x64 version of Visual Studio.

**Comment by lynnewu** [comment]
Just to be clear that I understand.  I need to design a user interface for my application, which doesn't exist yet.  

In order to design the user interface in anything other than the most trivial fashion, I have to run the application.  The application that doesn't really exist, but it's ok because I can see the UI change if I change the properties in Live View.

So...VBA has a designer view. WPF has/had a designer view.  WinForms has a designer view. VS 2022 *used* have a designer view, and you...removed it?  Why on earth did you do that?  Are you really trying to raise the barrier to entry for new developers?  Even in the html/js world I can get a live preview in a browser, and you're seriously asking me to spend this insane amount of money on VS Ent and I don't even get somethat that existed in freakin' WinForms?

Yeah, no.  Yes, I understand I will learn it, but if my choice is a learning cliff or just dealing with WinForms and HTML/JS, I know I'm going to select the side where I don't lose months and months of productivity just on the slim chance that enough people will want to use the same app on multiple applications, when I could just give them a web-based app much, much sooner.

This was a terrible decision.  There is no way I can go to my management and say, "Hey, you know how you wanted our apps to run everywhere?  Well, guess what?  Our UI devs are going to be in learning Hell for about a year.  You don't mind, right?  It's just money."

No.

**Comment by webluke** [comment]
I understand the push to have developers write the UI directly in code as we do with HTHM to create a cleaner UI XAML. But this is different. When developing UI, I spend 95%+ of my time in the XAML but having the live interpreted view of a page or display that is new and not accessible in the application is necessary for development. I can give up the poorly implemented drag-and-drop editing that has been a source of poor performance and bad XAML. But give us a preview of the results of our code and allow basic things like clicking on an element and bringing us to the code. The reason VISUAL Studio became such a success and made Windows a success was the ability to create UIs quickly and then link up the UI with data. 

If MAUI is to be the next cross-platform C# UI, it must be easy for new and old developers to create their apps, and the need for a live-in editor view of the UI is turning people away. It is also unfortunate that we see little to no response from Microsoft about this subject. Almost like a high-level manager made the decision, and no one wants to address the issue.

If anyone over at JetBrains is looking for a way to make more sales, adding a visual editor for MAUI XAML in Rider could be a selling point if they want to get more of a foot in the MAUI game.

**Comment by da-vics** [comment]
> I understand the push to have developers write the UI directly in code as we do with HTHM to create a cleaner UI XAML. But this is different. When developing UI, I spend 95%+ of my time in the XAML but having the live interpreted view of a page or display that is new and not accessible in the application is necessary for development. I can give up the poorly implemented drag-and-drop editing that has been a source of poor performance and bad XAML. But give us a preview of the results of our code and allow basic things like clicking on an element and bringing us to the code. The reason VISUAL Studio became such a success and made Windows a success was the ability to create UIs quickly and then link up the UI with data.
> 
> If MAUI is to be the next cross-platform C# UI, it must be easy for new and old developers to create their apps, and the need for a live-in editor view of the UI is turning people away. It is also unfortunate that we see little to no response from Microsoft about this subject. Almost like a high-level manager made the decision, and no one wants to address the issue.
> 
> If anyone over at JetBrains is looking for a way to make more sales, adding a visual editor for MAUI XAML in Rider could be a selling point if they want to get more of a foot in the MAUI game.

100% agree on this!

**Comment by davepruitt** [comment]
> I understand the push to have developers write the UI directly in code as we do with HTHM to create a cleaner UI XAML. 

I don't really understand the push to have developers write the UI directly in code. Let's take Flutter, for example. I actually really like Flutter. I think Google has done a great job with it. But my one _major_ complaint is how the UI is coded. It ends up becoming a massive file of spaghetti code, indented 10 million times and filled with 100 million parentheses. I find it really difficult to read. Unfortunately, a lot of modern tools are moving towards that same kind of UI design - one major one is React. But to me it just looks messy and unorganized.

Meanwhile, I love XAML. Being able to describe the UI in an XML-like format separate from the code logic really appeals to me. I don't need a drag-and-drop editor. I could care less about drag-and-drop functionality. But XAML really takes the cake compared to make-your-UI-in-code solutions. It's clean and intuitive.

Of course, having a XAML designer/previewer would really be nice! I like seeing the results of my XAML as I design it. I don't want to wait until I run the program and then navigate to the appropriate page to see the results of my XAML. The whole point of the XAML designer/previewer is so I can see how my UI looks as I design it. So, Microsoft, I would love to have a XAML designer/previewer. But drag-and-drop doesn't matter to me so much.

**Comment by da-vics** [comment]
> > I understand the push to have developers write the UI directly in code as we do with HTHM to create a cleaner UI XAML.
> 
> I don't really understand the push to have developers write the UI directly in code. Let's take Flutter, for example. I actually really like Flutter. I think Google has done a great job with it. But my one _major_ complaint is how the UI is coded. It ends up becoming a massive file of spaghetti code, indented 10 million times and filled with 100 million parentheses. I find it really difficult to read. Unfortunately, a lot of modern tools are moving towards that same kind of UI design - one major one is React. But to me it just looks messy and unorganized.
> 
> Meanwhile, I love XAML. Being able to describe the UI in an XML-like format separate from the code logic really appeals to me. I don't need a drag-and-drop editor. I could care less about drag-and-drop functionality. But XAML really takes the cake compared to make-your-UI-in-code solutions. It's clean and intuitive.
> 
> Of course, having a XAML designer/previewer would really be nice! I like seeing the results of my XAML as I design it. I don't want to wait until I run the program and then navigate to the appropriate page to see the results of my XAML. The whole point of the XAML designer/previewer is so I can see how my UI looks as I design it. So, Microsoft, I would love to have a XAML designer/previewer. But drag-and-drop doesn't matter to me so much.

Definitely! we shouldn't have to launch a simulator or connect a physical device to see UI changes. Especially Xaml.

**Comment by ArkTarusov** [comment]
The most disgusting experience of my life is trying to develop an application without seeing a visual result.
It is not clear how the GUI framework could in principle be released without previewing the layouts view.

**Comment by mcpontorez** [comment]
Please return or add XAML Designer like WPF, UWP in all UI XAML Products!

**Comment by nedoboy** [comment]
Developing without preview is really a painful process. We definitely need a XAML viewer!

**Comment by da-vics** [comment]
@jfversluis :)

**Comment by GuildOfCalamity** [comment]
I don't think they've ever provided a reason for why it was removed; I'm guessing because of the move to x-plat... but you could still provide a designer and just warn that "**visual style subject to change based on platform**". Maybe they could dust off **XAML Studio**? I'm reluctant to move away from WPF because of the lack of a real-time designer.

**Comment by davepruitt** [comment]
> I don't think they've ever provided a reason for why it was removed; I'm guessing because of the move to x-plat... but you could still provide a designer and just warn that "**visual style subject to change based on platform**". Maybe they could dust off **XAML Studio**? I'm reluctant to move away from WPF because of the lack of a real-time designer.

I remember reading somewhere (but I can't find it now) that the lack of a designer/previewer was partially motivated by some results from a focus group that they did. But if you look at the number of issues that have been created regarding this topic on the MAUI, Xamarin, and WinUI repositories, it is obvious their "focus group" was flawed.

**Comment by dragnilar** [comment]
> > I don't think they've ever provided a reason for why it was removed; I'm guessing because of the move to x-plat... but you could still provide a designer and just warn that "**visual style subject to change based on platform**". Maybe they could dust off **XAML Studio**? I'm reluctant to move away from WPF because of the lack of a real-time designer.
> 
> I remember reading somewhere (but I can't find it now) that the lack of a designer/previewer was partially motivated by some results from a focus group that they did. But if you look at the number of issues that have been created regarding this topic on the MAUI, Xamarin, and WinUI repositories, it is obvious their "focus group" was flawed.

I recollect the MAUI team giving a different excuse on a podcast/stream (it's on YT, I am having trouble finding the link right now). If I remember correctly, they said that they dropped support for the designer because they don't have anyone that can create a reasonable display that can match all MAUI's target platform/resolutions. They conceded that they would like to have it back in one day. 

I am just guessing, but I think we can both agree that there is more to this story. 

🤔💭

**Comment by k3ldar** [comment]
To have it one a single platform would be a bonus right now, you can always
extend support for it later

**Comment by GuildOfCalamity** [comment]
And why they can't enable the designer for `WinUI` projects at least? That still baffles me.

**Comment by jaywha** [comment]
I'd take a simple preview at the very least, y'know? Sure, it'd be nice to have the drag-'n-drop designer back too, but it should be pretty straightforward to just loosely interpret the XAML and ignore native look and feel, at first.  

WPF has been my bread and butter since my first job after college in 2017. I've got quite a few personal projects that are still in WPF and I'm _very_ accustomed to the workflow. If it's just not having someone to work on it, then I see a slew of interested community members that could make an extension possibly?

**Comment by LucNanquette** [comment]
A design view (like XAML/WPF) is an absolute must. I wonder what the "visual" in Visual Studio should else stand for?

**Comment by dragnilar** [comment]
> And why they can't enable the designer for `WinUI` projects at least? That still baffles me.

Agreed, it is baffling. The lack of a sincere and convincing explanation from MSFT unfortunately just leaves us to assume the worst.

**Comment by PhotoSoft** [comment]
I've programing WinForms app for years. I want to learn MAUI, but the lack of XAML designer is a big problem for me...

**Comment by eidylon** [comment]
Adding my hat to the rack, for what it's worth! Don't even need a designer per se, just something where I can see what my XAML will look like without needing to freaking run the application every time --- especially since device apps, on emulators, are QUITE slow to load. 🤦🏻‍♂️

**Comment by MrSyntax** [comment]
ridiculous how this feature is gone. I've been developing for 15+ years with Visual Studio. 
This is a step back to maybe even 20 years ago, and its unacceptable.
All i'm hearing and reading is excuses that dont hold ground.

As long as this feature isn't included, i can't recommend Visual Studio to anyone for cross-platform development.
Furthermore, because my team HAS TO develop multi platform, we'll as such most likely move away from Visual Studio alltogether. 

It makes no sense to stick to an IDE that actively obstructs and sabotages developers.

The only reason i can think of that Microsoft MAUI team didn't include this feature is because they actually DON'T want multi-platform applications. (probs cuz they scared ppl will move away from Windows?)

**Comment by davepruitt** [comment]
Microsoft peoples: do you have _any_ information on this at all?

**Comment by davepruitt** [comment]
> @MrSyntax moving away from VS isn't going to do you any good on its own. VS isn't the problem, it's the MAUI team's support (or lack thereof). If you can't build cross platform apps with MAUI, you are going to need a new technology entirely. 
> 
> Also, accusing MAUI of being afraid of other platforms makes no sense. That's their whole purpose.
> 
> I think the real issue is that MAUI is highly underfunded and the designer takes a lot of resources.
> 
> For my current team, we've decided to use Blazor (and then MAUI Blazor) instead. 

Yeah when you compare various multi-platform projects such as Flutter or React Native with MAUI, it really seems like Microsoft is underfunding MAUI and considers it a side project.

**Comment by davepruitt** [comment]
Listened to the WinUI community call this morning: they say a XAML previewer/designer it's still "on the backlog". So not really coming any time soon, unfortunately.

**Comment by encodedLife** [comment]
I can say without a doubt that the XAML Designer in Visual Studio was an essential tool for me. Being able to visually design the user interface of an application while simultaneously editing the corresponding XAML code was a huge time-saver, allowing me to quickly iterate and fine-tune my designs.

However, with the recent removal of the XAML Designer from Visual Studio, I feel like a key tool has been taken away from me. While I can still manually edit the XAML code, this is a much more time-consuming and error-prone process than using the visual designer. Without the XAML Designer, I feel like I am taking a step back in terms of productivity and efficiency, which is a major concern for me as a developer.

Moreover, the XAML Designer provided a way for developers to easily work with complex UI components such as data grids, list views, and tree views, which can be quite difficult to create and manage through code alone. Without this visual representation, it will be much harder to create and maintain these types of UI components, leading to more frustration and less polished applications.

Overall, the removal of the XAML Designer is a big back step for WPF-development, and I hope that Microsoft will reconsider this decision and bring back this essential tool for Visual Studio users.

**Comment by encodedLife** [comment]
I can say without a doubt that the XAML Designer in Visual Studio was an essential tool for me. Being able to visually design the user interface of an application while simultaneously editing the corresponding XAML code was a huge time-saver, allowing me to quickly iterate and fine-tune my designs.

However, with the recent removal of the XAML Designer from Visual Studio, I feel like a key tool has been taken away from me. While I can still manually edit the XAML code, this is a much more time-consuming and error-prone process than using the visual designer. Without the XAML Designer, I feel like I am taking a step back in terms of productivity and efficiency, which is a major concern for me as a developer.

Moreover, the XAML Designer provided a way for developers to easily work with complex UI components such as data grids, list views, and tree views, which can be quite difficult to create and manage through code alone. Without this visual representation, it will be much harder to create and maintain these types of UI components, leading to more frustration and less polished applications.

Overall, the removal of the XAML Designer is a big back step for WPF-development in my opinion, and I hope that Microsoft will reconsider this decision and bring back this essential tool for Visual Studio users.

**Comment by sangeethnandakumar** [comment]
This feature is indeed required for MAUI mass adoption. I had very hard time using XAML

**Comment by Mateo327** [comment]
The UI designer is the main component of a tool that lets you design UIs. Please bring it back for WinUI 3 and MAUI apps. Without it I'm forced to imagine what the result of my poorly written XAML is going to look like.

**Comment by sangeethnandakumar** [comment]
I was a WinForms developer and loved that tech. 
As it's getting older and non-cross-platform, Today I decided to try MAUI and felt it was so bad as I can't even see the UI because of the lack of a designer. I don't know what to do.

The hot reload is nowhere near a designer, it is useful only useful for tweaking already developed UI. 

Is there any update from Microsoft on this? This is very disappointing
I hope they'll bring a designer with WinUI on Visual Studio for Windows

**Comment by FatCatTuxedo** [comment]
Please we need this. MAUI is pointless without it!

**Comment by jnpl95** [comment]
This is one of the most commented on issues and highly upvoted. Please Maui team, add this feature. Having to run the app to view changes via hot reload isn't the best way of rapidly prototyping ui elements. And I feel this is what is holding many developers back from using MAUI.

**Comment by MagicAndre1981** [comment]
Microsoft doesn't care about it. They created a [new item](https://github.com/dotnet/maui/issues/14032) to only track Hotreload improvements in .NET8:

> In .NET 8 we continue to pursue improvements to the reliability of Hot Reload

**Comment by dragnilar** [comment]
> Microsoft doesn't care about it. They created a [new item](https://github.com/dotnet/maui/issues/14032) to only track Hotreload improvements in .NET8:
> 
> > In .NET 8 we continue to pursue improvements to the reliability of Hot Reload

It isn't a surprise. I suppose to count our blessings, let's just be glad that they're trying to make Hot Reload "hotter". 

That being said, I strongly believe Hot Reload will never be able to support the same level of productivity that we could appreciate with an in-VS previewer and/or an actual designer.

**Comment by eidylon** [comment]
> That being said, I strongly believe Hot Reload will never be able to support the same level of productivity that we could appreciate with an in-VS previewer and/or an actual designer.

Especially since having to rebuild/reload/redeploy your app to the emulator takes a good 5-10 minutes! .... well, for Android dev anyway. Compared to being able to see your changes previewed in real-time in a designer. I just don't get how MS doesn't consider this a worthy improvement. 

Step 1. Click Reload 
Step 2. Go brew a new cup of coffee

**Comment by XeonG** [comment]
I just think MS are beyond all hope of ever getting better in anything..

MAUI.. a stupid product name for one.. it's a damn island ffs, search results on training material SEO anyone?  of which there is not much as development pace is slow.. yeah there are updates might be every couple of months but that's more of a hinderance as they change more than they add or improve so any tutorials are a broken waste of time. 

Then you got other frameworks like UNO and Avaloniaui .... doing the same thing really but also none really providing everything you'd want.

The direction and pace at which MS is going with any of this stuff that people do want is a joke.. and the lack of a design editor  just suggests they don't even think it's worth it.

And Blazor? Maui? Blazor? 🔥Mess Reckage

> "Winforms has one of the best drag and drop designers that Microsoft has ever produced."
> 
> That was clearly from a day and time when MS had employees who could actually make decent tools and put together consistent UI's.
> 
> Instead they buy crap like Xamarin. And there OS looks like it's been pulled apart and taken over by literal retards turning into absolute rubbish.. If it wasn't for the third party dev community putting up patches to fix it, like Explorer Patcher https://github.com/valinet/ExplorerPatcher I would have already left development on this trainwreck that is Windows.
> 
> And I dunno who made there Dark Mode theme for Win11 but it was like they hired a bunch of interns who just thought rounding some corners off was enough.. Leaving just about every winform GUI control completely unthemed while breaking support for anyone to fix it. So most UI you see in Windows is just white.. in a dark mode theme.. moron development.
> 
> Anyway I'll just stick to using JavaFX they have a Scenebuilder for quick prototyping and the build pipelines work for all platforms..ie Linux included.. Or use QT
> 
> MS direction with just about everything now is hair brained and inconsistent, about the consistent thing is it being constantly broken, barely fit for purpose and really unfinished rubbish.

**Comment by dragnilar** [comment]
@XeonG Funny you mentioned the Blazor + MAUI thing... someone on my team showed me a rather unpleasant rumor this morning:

https://www.reddit.com/r/csharp/comments/122yelz/maui_going_to_all_in_on_blazor_hybrid/

_IF_ this rumor has any truth to it, I could see that a designer in MAUI could end up going from a "it's a known need" to "it's never happening".

Disclaimer- I love Blazor but I am not convinced it is the answer to everything.

**Comment by davepruitt** [comment]
> @XeonG Funny you mentioned the Blazor + MAUI thing... someone on my team showed me a rather unpleasant rumor this morning:
> 
> https://www.reddit.com/r/csharp/comments/122yelz/maui_going_to_all_in_on_blazor_hybrid/
> 
> _IF_ this rumor has any truth to it, I could see that a XAML previewer/designer in MAUI could end up going from a "it's a known need" to "it's never happening".
> 
> Disclaimer- I love Blazor but I am not convinced it is the answer to everything.

This is incredibly disheartening and frustrating news, if it is true. I keep getting pushed more and more towards Flutter with every day that passes.

**Comment by da-vics** [comment]
> > @XeonG Funny you mentioned the Blazor + MAUI thing... someone on my team showed me a rather unpleasant rumor this morning:
> > https://www.reddit.com/r/csharp/comments/122yelz/maui_going_to_all_in_on_blazor_hybrid/
> > _IF_ this rumor has any truth to it, I could see that a XAML previewer/designer in MAUI could end up going from a "it's a known need" to "it's never happening".
> > Disclaimer- I love Blazor but I am not convinced it is the answer to everything.
> 
> This is incredibly disheartening and frustrating news, if it is true. I keep getting pushed more and more towards Flutter with every day that passes.

hmmm

**Comment by sangeethnandakumar** [comment]
Investing in a drag and drop "designer" will greatly reduces people's dissatisfaction towards MAUI rather than decommissioning the product I would say .!

**Comment by level54** [comment]
The fact that there's only 85 up votes on the original post for this thread, tells me that there's probably only about 102 developers in the world using, or attempting to use MAUI. So I'm going to give a wide berth until such time as Microsoft can decide what they want MAUI to be, because at the moment, it's a complete and utter waste of time.

**Comment by ToolmakerSteve** [comment]
@davepruitt - sadly, it does feel like someone high up at Microsoft is under the delusion that `Blazor` is their solution to "UI that runs everywhere".

On the one hand, its hard to argue with the business model of developing using web standards, and having that run everywhere.

But its sad (as a long-time developer of desktop and mobile apps) if app development becomes entangled with complications/limitations/techniques that are necessary on the web, but not optimal for apps.

**Comment by XeonG** [comment]
" UI Toolkit. (I haven't used it, just based on what I read.)"  ..its shit..and leaves a lot to be desired...  but then so is everything these days. 

kinda crazy how bad it has all gotten really.

**Comment by davepruitt** [comment]
Today I went to design a new pop-up that we are using in our app, and was reminded how useful it would be to have a designer, and how utterly insane and mindblowing it is that Microsoft has not yet created a designer for MAUI, given how much desire there is for this in the community.

Hot reload is not a good replacement for a XAML previewer/designer.

**Comment by GuildOfCalamity** [comment]
TBH I would use **Hot Reload**, if it worked properly... but it doesn't. How do designers at MS handle UI development?... are they really satisfied with the time wasted in adjusting the margin on a control and then having to build the solution again to see if it's correct?

**Comment by sangeethnandakumar** [comment]
@Shadowblitz16 @jsuarezruiz @Eilon - Please do something as a lot of people waiting to try MAUI needs this to effectively use this framework. WinForms was a huge success because of the innovative ways Microsoft did at that time.
Why keeping MAUI not appealing to most new devs who expect a designer?

**Comment by level54** [comment]
I hate to be the bearer of bad tidings, but MS has no intention of listening to our whining. They are more interested in concentrating on getting MAUI ready for .NET 8 (see [this](https://devblogs.microsoft.com/dotnet/announcing-dotnet-maui-in-dotnet-8-preview-6/))

And after that they'll be working on getting it ready for .NET 9 and so on and so forth. All/most of the bugs fixed in the latest preview was all for the desktop, ie Windows, so anybody using MAUI for devices will have to wait.....probably "at infinitum".

**Comment by dragnilar** [comment]
@level54 You're unfortunately right. It seems MSFT is probably on this course until Sataya is gone. I don't like blaming this all on the man, but I am under the impression he doesn't care about whether or not visual studio is "visual". So to me, his apathy has impacted the products. If someone else sees things differently, feel free to paint a brighter picture.

**Comment by davepruitt** [comment]
> I hate to be the bearer of bad tidings, but MS has no intention of listening to our whining. They are more interested in concentrating on getting MAUI ready for .NET 8 (see [this](https://devblogs.microsoft.com/dotnet/announcing-dotnet-maui-in-dotnet-8-preview-6/))
> 
> And after that they'll be working on getting it ready for .NET 9 and so on and so forth. All/most of the bugs fixed in the latest preview was all for the desktop, ie Windows, so anybody using MAUI for devices will have to wait.....probably "at infinitum".

Which just goes to show how underfunded/understaffed the team must be. If there has to be so much focus on "preparing MAUI for the next .NET version", nothing will ever actually get done.

Continuing my comment from yesterday: I spent 2-3 hours creating a pop-up dialog because I had to continuously compile/re-build/re-deploy every single time I made a change to the XAML. Then of course I had to also navigate to the correct place in my app to view the pop-up dialog. Hot reload didn't work at all. The whole process would have taken me 10 minutes if I had a XAML designer.

**Comment by XeonG** [comment]
"Continuing my comment from yesterday: I spent 2-3 hours creating a pop-up dialog because I had to continuously compile/re-build/re-deploy every single time I made a change to the XAML. "

Sounds terrible, by design this must be the goal of MS.. waste people's time, so long as you aren't ditching everything they do and still doing something with there crap they'll keep on track with this bullshit.

They don't want to invest in building such tools for there own stuff, which shows how much confidence they have in it. The fact MAUI (stupid product name 🏝) has the competition in the same space  with UNO/Avalonia just makes everything MS is doing in this space seem dumb, they have the money to fund doing something better.

**Comment by dragnilar** [comment]
> > I hate to be the bearer of bad tidings, but MS has no intention of listening to our whining. They are more interested in concentrating on getting MAUI ready for .NET 8 (see [this](https://devblogs.microsoft.com/dotnet/announcing-dotnet-maui-in-dotnet-8-preview-6/))
> > 
> > And after that they'll be working on getting it ready for .NET 9 and so on and so forth. All/most of the bugs fixed in the latest preview was all for the desktop, ie Windows, so anybody using MAUI for devices will have to wait.....probably "at infinitum".
> 
> Which just goes to show how underfunded/understaffed the team must be. If there has to be so much focus on "preparing MAUI for the next .NET version", nothing will ever actually get done.
> 
> Continuing my comment from yesterday: I spent 2-3 hours creating a pop-up dialog because I had to continuously compile/re-build/re-deploy every single time I made a change to the XAML. Then of course I had to also navigate to the correct place in my app to view the pop-up dialog. Hot reload didn't work at all. The whole process would have taken me 10 minutes if I had a XAML designer.

I believe it. It's even worse imo if you have Android involved.

**Comment by soroshsabz** [comment]
ITNOA

I think WinUI 3 and MAUI has same infrastructure for XAML Islands, so I think Microsoft must consider all request from [WinUI 3.0 XAML Designer](https://github.com/microsoft/microsoft-ui-xaml/issues/5917) too for prioritizing this request to implement or not.

**Comment by Noemata** [comment]
Name any products released by Microsoft that use either MAUI or WinUI.  Only UWP had a brief period of adoption by Microsoft during the Windows 8/10 life cycle.  If Microsoft makes extremely limited use of MAUI and WinUI for their own development, why should you use it at all?  That's the way I'm looking at the current state of Microsoft tech.  Only invest in tech actually used internally by Microsoft.  The roadmaps and declarations of support were very promising.  At this page -> https://dotnet.microsoft.com/en-us/platform/support/policy/maui , tvOS is listed as supported by MAUI.  It is not!  It's these sorts of unfulfilled promises that lead folks astray.  WinUI was supposed to be open sourced two years ago.  Still waiting.  WinUI was supposed to support the Map control and inking (both prominently declared on the original roadmap).  Microsoft silently stopped mentioning support for the Map control, Inking, and a half dozen other critical items like validation.  Just look at the sad state of UWP today.  It's hard to find any official mention of it.

And for those not in the know, the WinUI team is mostly disbanded.  As goes WPF, UWP, Xamarin, WinUI so goes MAUI before long.

Bitter.  Yes, very.

**Comment by Enxyphered** [comment]
Absolutely disgusting that they think it's okay to try and force people to use MAUI without a previewer that doesn't require us to spin up the application. Not only that we have to code in navigation to the page that we would like before we can even start seeing what we are editing. Beyond scum. Absolutely incompetency at it's best.

**Comment by vslee** [comment]
@Enxyphered, I am interested in a MAUI designer as much as anyone else, but let's keep things respectful. Hurling insults isn't helpful.

**Comment by CAKCy** [comment]
Please read and act accordingly!

![image](https://github.com/microsoft/microsoft-ui-xaml/assets/7456966/7f950b63-7462-413d-9611-734eee4c48fc)

The link again:
https://developercommunity.visualstudio.com/t/Feature-request-Visual-designer-for-Wi/1608476?q=WinUI+Designer

Oh! And SPREAD THE WORD! 
PLEASE!

**Comment by davepruitt** [comment]
I went to the link you provided and I up-voted it and posted a comment. Once again I am so amazed that Microsoft has basically ignored this issue for _years_ now.

**Comment by MSBassSinger** [comment]
The lack of drag-and-drop, visual designers like the WinForms designer is a worse problem for MS than they admit to realizing.

Consider that in the 1990s, one small team created the first (and wildly popular and emulated by competitors) visual designer for Visual Basic.  Over time, this was migrated to Visual Studio when it replaced the Visual Basic IDE and Visual C++ IDE.  It was popular precisely because it added value that "hot reload" designers cannot.

How?

- For any given project, there is a set deadline by which to deliver.  
- There are finite development resources for the code writing portion of the SDLC, generally counted as number of developers.
- The number of developers have a limited number of person-hours each to finish their work.
- The person hours per developer can be calculated as T = #HoursUntilDelivery/#Dev 
- Without a drag-and-drop designer, a significantly larger percent of T is spent on UI design.
- Likewise, a significantly smaller percent of T is devoted to the code behind (validation, business logic, data access, etc.).
- If the development phase of the SDLC were an iceberg, the code behind portion is the part underwater (which is the vast majority).
- With a drag-and-drop designer, a much smaller percent of T is spent on UI design.
- Likewise, a significantly larger percentage of T is devoted to code behind, thus making the time spent proportional to the need and improving the quality of the project.  That means less QA time is needed, and less refactoring initiated from the user testing.
- Microsoft developers and program managers have come up through the command line/scripting paradigm that Windows developers left behind in the mid-1990s. To them, "hot reload" is a game changer, but only in the sense that someone from 1850 would see a rotary phone as a technological leap, ignoring the far greater productivity and ease of a 2023 smartphone.
- As has happened throughout software development history, if Microsoft is unable to get their heads out of the rotary phone age and into the smartphone age (so to speak), they will go the way of those Microsoft replaced in past years that refused to make wise choices.  Where is dBase, Wordperfect, Lotus 1-2-3, and others that either disappeared or exist only with a tiny number of users?  Microsoft is not immune to that, and if Visual Studio is abandoned for UI development because the MS program managers are too enthralled with rotary phones, then it is only a matter of time before other MS products follow suit.

There is a solid argument, both in terms of revenue and technology, for a single drag-and-drop UI builder that services WinForms, MAUI (XAML), Blazor (HTML), and perhaps others by using the same principle that .NET uses when it writes multiple coding languages (C#, VB, etc.) as the same MSIL code. But first, MS has to assign or hire those with the vision and intellectual capabilities to do what Alan Cooper and his small team did (in principle) for WinForms back in the 1990s.

**Comment by MSBassSinger** [comment]
Please consider upvotes for:
https://developercommunity.visualstudio.com/t/Create-Common-Drag-and-drop-Designer-for/10501474

**Comment by soroshsabz** [comment]
> Please consider upvotes for:
> https://developercommunity.visualstudio.com/t/Create-Common-Drag-and-drop-Designer-for/10501474

@MSBassSinger please add your great description in to the main developer community ticket for bring back visual designer for all UI frameworks

Thanks

**Comment by soroshsabz** [comment]
@MSBassSinger 

Microsoft change [Developer Community ticket](https://developercommunity.visualstudio.com/t/XAML-Designer-for-Net-MAUI/10224319) for this request, so please vote up this ticket.

https://developercommunity.visualstudio.com/t/XAML-Designer-for-Net-MAUI/10224319

**Comment by pinkfloyd11** [comment]
Please add Visual Designer for MAUI.
It is nonsense not have it in 2024...

**Comment by Daniel-Sh-uq** [comment]
At this point if MS doesn't want to give us at least normal XAML preview, lets make MAUI support html: https://github.com/dotnet/maui/issues/290. 
Because developing MAUI without proper preview is just killing the productivity.

**Comment by dy-dev** [comment]
Has anyone switched to Avalonia Framework in order to have a working viewer?
I'd be happy to have feedback on how things went

**Comment by GuildOfCalamity** [comment]
@dy-dev I tried avalonia for a while but it was clunky to use. I'd go back to WPF until someone at Microsoft can take the rose-colored glasses off and add a designer to WinUI3.

**Comment by vamman** [comment]
When does Microsoft release the internal version of the Maui Visual Editor?

**Comment by GMPrakhar** [comment]
I was newly trying out MAUI and stumbled upon this. I have created a repo to handle the designer task myself: https://github.com/GMPrakhar/MAUI-Designer. Currently it supports the rendering of the XAML live, e.g:

![image](https://github.com/user-attachments/assets/09dd2c10-f81c-4dea-bc98-32a63e7023ce)
![image](https://github.com/user-attachments/assets/c2f690dc-45bc-40c1-9bf3-18f731fd5729)
![image](https://github.com/user-attachments/assets/c4aa0369-862a-47a9-bf16-bc6426a3d877)

It was quite easy, as MAUI has an extension method that supports direct XAML rendering. Not sure why isn't it being used yet?

**Comment by soroshsabz** [comment]
@samhouts What is meaning of changing milestones? maybe we can be happy to Microsoft wants to start development of UI Designer?

**Comment by sasakrsmanovic** [comment]
Many in this thread will be pleased to learn that a Visual UI Designer now exists for cross-platform .NET Apps. It is called Hot Design, provided via Uno Platform, so you are still using .NET, Visual Studio (or other IDEs) and and your existing skills to get the work done. 

Announcement: https://github.com/unoplatform/uno/discussions/18833

.NET Conf 2024 Session on Hot Design:
https://www.youtube.com/watch?v=sJPyieyt1Rc&list=PLdo4fOcmZ0oXeSG8BgCVru3zQtw_K4ANY&index=93

**Comment by Supergeek** [comment]
Hot Design is a paid feature.

**Comment by XeonG** [comment]
Crazy a third party has made a visual designer that is only half as good as winforms then stuck it behind a paywall, while MS had a winform designer for free and now can't even bothered to make anything let alone something even half as good.

**Comment by jnpl95** [comment]
Microsoft closed the feedback thread a couple days ago
https://developercommunity.visualstudio.com/t/XAML-Designer-for-Net-MAUI/10224319

Microsoft responded with this:
"Microsoft Resolution - 
Closed - Out of Scope
Thank you for the suggestion. We appreciate everyone’s feedback and understand the need for an efficient UI design experience. A drag-and-drop UI designer is not part of our direction for .NET MAUI. We’re focused on providing a high-fidelity design-time experience through XAML Live Preview and Hot Reload, which reflects real-time changes using the actual rendering pipeline. This approach ensures consistency across device platforms and aligns with modern development workflows. You should be able to efficiently create and iterate on your UI, particularly when paired with Copilot Vision.

To make this experience better, XAML Live Preview is available at design-time in Visual Studio 2022 version 17.14 and Visual Studio 2026."

So yeah, not going to happen unfortunately

**Comment by XeonG** [comment]
https://opensilver.net/whats-new/  and  https://xaml.io       ....I found recently.

 "You should be able to efficiently create and iterate on your UI, particularly when paired with Copilot Vision"   yeah good job MS, maybe copilot can just make the UI designer aswel yeah

**Comment by jfversluis** [comment]
As indicated in the VS Feedback ticket, this is not going to happen for the foreseeable future, we don't have much to add from our side. While related to .NET MAUI this is ultimately a tooling/Visual Studio feature. Thank you for all the feedback and discussion!
